ArjunaJTA/jms/src/main/java/org/jboss/narayana/jta/jms/JmsXAResourceRecoveryHelper.java

```diff
@@ -82,31 +71,31 @@ public boolean initialise(String properties) {
      */
     @Override
     public XAResource[] getXAResources() {
-        if (connect()) {
-            if (jtaLogger.logger.isTraceEnabled()) {
-                jtaLogger.logger.trace("Returning XA resource: " + this);
+        if (!connectionManager.isConnected()) {
+            try {
+                connectionManager.connect();
+            } catch (XAException ignored) {
+                return new XAResource[0];
             }
-
-            return new XAResource[] { this };
         }
 
-        return new XAResource[0];
+        return new XAResource[] { this };
     }
 
     /**
      * Delegates XAResource#recover call to the connected JMS resource. If provided argument is XAResource.TMENDRSCAN, then JMS
      * connection will be closed at the end of the call.
      *
-     * @param i
+     * @param flag
      * @throws XAException
      */
     @Override
-    public Xid[] recover(int i) throws XAException {
+    public Xid[] recover(int flag) throws XAException {
         try {
-            return delegate.recover(i);
+            return connectionManager.connectAndApply(delegate -> delegate.recover(flag));
         } finally {
-            if (i == XAResource.TMENDRSCAN) {
-                disconnect();
+            if (flag == XAResource.TMENDRSCAN) {
+                connectionManager.disconnect();
```

Comment:
Sorry for a delay, I was away last week.
@mmusgrov it used to be closed at TMENDRSCAN before, so nothing has changed.
@tomjenkinson, we did discuss it the first time I've implemented this and we end up deciding to open connection on getXAResources and closing it on TMENDRSCAN. I'm fine with keeping connection open as long as it is an acceptable practice. It would make implementation easier too.